lib/chefspec/api/stubs_for.rb

```diff
@@ -0,0 +1,134 @@
+require "chef/version"
+
+module ChefSpec
+  module API
+    module StubsFor
+      # Pull in the needed machinery to use `before` here.
+      extend RSpec::SharedContext
+
+      # Which version to use the shell_out_compacted hook on.
+      HAS_SHELLOUT_COMPACTED = Gem::Requirement.create("> 14.2")
+
+      # Hook used in the monkey patches to set up a place to inject stubs when
+      # needed for a resource or provider.
+      #
+      # @api private
+      # @param object [Chef::Resource, Chef::Provider] Resource or provider to inject
+      # @param type [Symbol] Type of object to register stubs on
+      # @return [void]
+      def self.setup_stubs_for(object, type)
+        # This space left intentionally blank, real implementation is below.
+      end
+
+      # Place to hold any stubs which should be active for this example.
+      let(:_chefspec_stubs_for_registry) do
+        # Auto-vivify for things like _chefspec_stubs_for_registry[:resource]['my_resource'] == [].
+        Hash.new do |hash, key|
+          hash[key] = Hash.new do |inner_hash, inner_key|
+            inner_hash[inner_key] = []
+          end
+        end
+      end
+
+      # Take over from the default, no-op implementation of {.setup_stubs_for}.
+      before do
+        allow(StubsFor).to receive(:setup_stubs_for) do |object, type|
+          type_stubs = _chefspec_stubs_for_registry[type]
+          resource_object = object.is_a?(Chef::Provider) ? object.new_resource : object
+          stubs = type_stubs[nil] + type_stubs[resource_object.resource_name.to_s] + type_stubs[resource_object.to_s]
+          stubs.each do |stub|
+            instance_exec(object, &stub)
+          end
+        end
+      end
+
+      # Register stubs for resource objects.
```

Comment:
maybe it would be good to list where in custom resources the stubs are like calls inside `#load_current_value` etc.

Reply:
Yeah, that will need to be in the reference docs since it definitely gets funky if you don't grok the internals.